Templating feature

adrianhurt opened the pull request.

adrianhurt commented:
Hi, this PR allows you to define your own template (and not only the DOT one) to generate an output file (or a set of files) from your information on crane.yml.

If you agree with this, I will add the tests and docs. But for the moment is a draft.

As I said, for the moment I haven't documented it, but in summary, it's a generalized feature for the current graph option, but with your own template. There are 2 arguments:

* Input template file (`-t`, `--tmpl`). Required.
* Output file (`-o`, `--output`). Optional. If it's not present, the output will be printed in the console (`os.Stdout`). If it's present there are 2 options:
  * Regular file name (ex: _myGraph.dot_): the output will be saved in that file.
  * File format containing `%s` (ex: _%s.service_): it will generate a file for each container applying the same template for all of them. This format determines the name of each file.

I've included some template examples for _DOT graph_, _Systemd_ and _Upstart_. In my case I've used the _Systemd_ example in _CoreOS_ and works perfectly taking into account all the dependencies and generating every _service_ file.

To use the examples:
```
crane template -t path/to/graph.tmpl -o output/myGraph.dot
crane template -t path/to/systemd.tmpl -o output/%s.service
crane template -t path/to/upstart.tmpl -o output/%s.conf
```

bjaglin commented on crane/template.go:
```diff
@@ -0,0 +1,105 @@
+type TemplateDataForContainer struct {
+	Graph              DependencyGraph
+	TargetedContainers Containers
+	CurrentContainer   Container
```
I guess we should reduce the amount of stuff exposed in `Container` - side-effect functions like `Run()` or `Kill()` should not be available.

adrianhurt replied:
Yes! maybe called `ContainerData`?
```
type ContainerData interface {
	Name() string
	Dockerfile() string
	Image() string
	Id() string
	Dependencies() *Dependencies
	Exists() bool
	Running() bool
	Paused() bool
	ImageExists() bool
	Status() []string
}
```

bjaglin commented on crane/template.go:
```diff
@@ -0,0 +1,105 @@
+	TargetedContainers Containers
+}
+type TemplateDataForContainer struct {
+	Graph              DependencyGraph
```
might be worth considering exposing `Dependencies` as an interface rather than a struct for the sake of abstraction

bjaglin replied:
Also, I guess the original `Graph`/`DependencyGraph` naming is a bit off, since it's a structure representing a graph (which happens to be a tree) but expressed "flat".

adrianhurt replied:
For `Dependencies`, do you mean to create an interface mapping the current struct? Called `Dependencies` as well?

And for `Graph`, I think I understand your point… maybe `DependencyMap`?

bjaglin commented on crane/cmd.go:
```diff
@@ -201,7 +211,7 @@ in the config, but not defined). Targeted containers are highlighted with color
 net=container relations.`,
 		Run: configCommand(func(config Config) {
-			config.DependencyGraph().DOT(os.Stdout, config.TargetedContainers())
+			ConfigToTemplateData(config).DOT(options.tmplOutput)
```
I guess you can take out the old function in DependencyGraph() then

adrianhurt replied:
Yes, of course. I intended avoid to touch the rest of the code for the draft.

bjaglin commented on crane/cmd.go:
```diff
@@ -252,6 +262,11 @@ See the corresponding docker commands for more information.`,
 	cmdLogs.Flags().StringVarP(&options.tail, "tail", "", "all", "Output the specified number of lines at the end of logs")
 	cmdLogs.Flags().BoolVarP(&options.colorize, "colorize", "z", false, "Output the lines with one color per container")
 
+	cmdTmpl.Flags().StringVarP(&options.tmplInput, "tmpl", "t", "", "The input template file")
```
why not spell `template` out completely?

adrianhurt replied:
Ok, it was simply for avoiding `crane template —template "file.tmpl"`. It seems too redundant.

bjaglin commented on crane/cmd.go:
```diff
@@ -252,6 +262,11 @@ See the corresponding docker commands for more information.`,
 	cmdLogs.Flags().BoolVarP(&options.colorize, "colorize", "z", false, "Output the lines with one color per container")
 
+	cmdTmpl.Flags().StringVarP(&options.tmplInput, "tmpl", "t", "", "The input template file")
+	cmdTmpl.Flags().StringVarP(&options.tmplOutput, "output", "o", "", "The output file or directory (os.Stdout as default)")
```
how would directory work?

adrianhurt replied:
Oh, ignore that, sorry. It's because previously I thought in indicating a directory to output every file with the name of each container, but then I decided to use a string format with `%s` to be able to adding a file extension. I will check every text, I only was waiting to the PR is approved.

bjaglin commented:
:+1: I personally love it, great idea! I guess we just need to agree on what's available in the template - the current code isn't particularly thought for being used externally.

adrianhurt commented:
Thanks! And yes, the idea it was already there for DOT graphs. It's only a little refactoring to do it more general.

adrianhurt commented:
BTW, I added some utility functions for the templates. I chose them because I think they could be the most useful ones, but maybe I forgot someone else. There are:
* strings.Contains
* strings.ContainsAny
* strings.HasPrefix
* strings.Join
* strings.Split
* strings.ToLower
* strings.ToTitle
* strings.ToUpper
* regexp.MatchString

bjaglin commented on README.md:
````diff
@@ -192,6 +195,74 @@ This could be used like so: `crane provision service1`, `crane run -v databases`
+The data object passed to the template is a `TemplateInfo` struct, or a `TemplateInfoForContainer` one if you want to generate a file per container. You can see here the structure of each struct or interface needed.
+
+```go
+type TemplateInfo struct {
````
This might get outdated - a link to https://godoc.org/github.com/michaelsauter/crane/crane might be better to avoid keeping that in sync?

adrianhurt replied:
Yes, I agree. Particularly https://godoc.org/github.com/michaelsauter/crane/crane#TemplateInfo and https://godoc.org/github.com/michaelsauter/crane/crane#TemplateInfoForContainer

bjaglin commented on crane/test_templates/test.tmpl:
```diff
@@ -0,0 +1,17 @@
+    imageExists: {{ .ImageExists }}
+    status: [{{ Join .Status ", " }}]{{ end }}{{ end }}
+
+DEPENDECY-MAP:{{ range $name, $dependencies := .DependencyMap }}
```
typo - `DEPENDENCY-MAP`

adrianhurt replied:
Thanks

bjaglin commented on crane/template.go:
```diff
@@ -0,0 +1,131 @@
+}
+
+// generates a file with the output. If outputPath is empty, it will be dumped to os.Stdout
+func generateFile(tmpl *template.Template, outputPath string, templateInfo interface{}) error {
```
no client seems to care about that error returned

adrianhurt replied:
I had 
```
err = generateFile(tmpl, outputPath, templateInfo)
if err != nil {
  return
}
```
But at the end of the function it has no effect in that point. And the error is already printed inside the function

bjaglin replied:
then don't return an error at all? but shouldn't an error be fatal at this point?

adrianhurt replied:
with a `panic(StatusError{fmt.Errorf(errorDescription, errorStatus})`? Which `errorStatus` can I use? There is any convention?

bjaglin commented on crane/cmd.go:
```diff
@@ -252,6 +268,11 @@ See the corresponding docker commands for more information.`,
 	cmdLogs.Flags().StringVarP(&options.tail, "tail", "", "all", "Output the specified number of lines at the end of logs")
 	cmdLogs.Flags().BoolVarP(&options.colorize, "colorize", "z", false, "Output the lines with one color per container")
 
+	cmdTmpl.Flags().StringVarP(&options.tmplInput, "template", "t", "", "The input template file name")
```
continuing here the discussion about `template` as the long parameter name... I see your point about the fact that it's weird to have both the command and a flag with a similar key... Maybe that calls for a different command name - `generate`, `transform` ?

adrianhurt replied:
I see your point. Now I'm not sure which name should have the feature… With `template` I understand better what it is intend, and the other ones are more ambiguous. But the action itself is not template but the file used to generate an output. So I would prefer `generate` for this point. The problem I see for using `generate` is that it's a very generic word.

bjaglin commented on crane/template.go:
```diff
@@ -0,0 +1,131 @@
+	"text/template"
+)
+
+type TemplateInfo struct {
```
Maybe I am too defensive and internal interfaces shouldn't be public as they are now, but I feel all exposed structures should be subset interfaces of the internal ones - so it's clear that we break the external contract if we remove/update an internal function/struct/whatever.

adrianhurt replied:
Maybe

bjaglin commented on crane/template.go:
```diff
@@ -0,0 +1,131 @@
+// generates a textual output for the selected template and templateInfo
+func (templateInfo TemplateInfo) template(templatePath string, outputPath string) {
+	if templatePath == "" {
+		print.Errorf("You must select the input template with the parameter --template (-t)\n")
```
shouldn't that be enforced earlier? It feels weird to see hardcoded CLI stuff here

adrianhurt replied:
Yes, you're right. In deed, it was a question I would like to ask you and I forgot. How can I force a flag to be present (with Cobra)?

bjaglin replied:
I don't think you can, you'd have to add a checker function called from all command handlers utilizing this option. Not perfect, but as long as it's contained together with the other CLI stuff, it's not too ugly.

adrianhurt replied:
Something like this?
```
func requiredStringFlag(name string, value string) {
  if value == "" {
    panic(StatusError{fmt.Errorf("The flag '%s' is required", name), XX})  //  XX = 64??
  }
}
...
Run: configCommand(func(config Config) {
  requiredStringFlag("template", options.tmplInput)
  ConfigToTemplateInfo(config).template(options.tmplInput, options.tmplOutput)
}, true),
...
```

bjaglin commented on crane/cmd.go:
```diff
@@ -192,6 +193,21 @@ formatted in bold provided the 'colorize' flag is on.`,
+		Long: `Generates a textual output using Go templates. Use the 'template' flag to
+indicate the input template. Depending on the optional 'output' flag's value the
+behaviour is different. If you leave it, the output will be dumped through Stdout.
+If you specify a filename, the output will be rendered on that file. Additionally,
```
`Additionally` makes me think that `the output will be rendered on that file` remains true in the `%s` case, which is not the case, it really is an either-or.

adrianhurt replied:
Right, then `Additionally` -> `Finally`?

bjaglin commented on templates/readme.md:
```diff
@@ -0,0 +1,61 @@
+
+Generate a [DOT](http://en.wikipedia.org/wiki/DOT_(graph_description_language)) file representing the dependency graph. Bold nodes represent the containers declared in the config (as opposed to non-bold ones that are referenced in the config, but not defined). Targeted containers are highlighted with color borders. Solid edges represent links, dashed edges volumesFrom, and dotted edges net=contai ...
+
+# Systemd template
```
couldn't these templates be tested from `template_test.go` to make sure they remain valid no matter what we do (horrible conscious breaking changed or other) ?

adrianhurt replied:
Yes, but I think the idea should be that these are only examples. Crane shouldn't support them… I can add more templates in the `test_templates` folder for them.

bjaglin replied:
I understood this was given as an example/suggestion, but I just hate to see documentation examples which don't work, so running them as part of the automated testing would prevent that.

bjaglin commented on crane/template.go:
```diff
@@ -0,0 +1,131 @@
+	Dockerfile() string
+	Image() string
+	Id() string
+	Dependencies() *Dependencies
```
I know this makes life easier when generating a per-container systemd/upstart file, but I feel like this shouldn't be exposed as-is here. I see `ContainerInfo` as something providing either raw configuration, or state about that container. `Dependencies` is an opinionated transformation of  links/volumesFrom/net, and is already available via a lookup in the `DependencyMap`. If we need access to raw dependencies, maybe we could expose `RunParams()` ? I know it seems clunky but what if we need access to the exposed ports from the template?

adrianhurt replied:
I see your point with this as well. We could remove `Dependencies` from `ContainerInfo` and check them directly from `DependencyMap`.

For the other hand, it's true that making `RunParams()` accessible could be interesting. But then `RmParams()` and `StartParams()` as well?

In fact, I was thinking to give all the information of `config` struct, but interface `Config` limit us now, so I would need to add some methods. I mean to have something like this:
```
type TemplateInfo struct {
  ContainerMap   map[string]ContainerInfo
  DependencyMap  DependencyGraph
  Target         []string
  Groups         map[string][]string
}

func (ti TemplateInfo) TargetedContainers() []ContainerInfo
func (ti TemplateInfo) Dependencies(c ContainerInfo) *Dependencies
func (ti TemplateInfo) ContainersForGroup(g string) []ContainerInfo
func (ti TemplateInfo) IsTarget(c ContainerInfo) bool
func (ti TemplateInfo) BelongsToGroup(g string) bool
// you can think more...
```

bjaglin commented:
Just a heads up regarding my comments - @michaelsauter still hasn't given his opinion about whether/ how should go in, so take them with a grain of salt.

adrianhurt commented:
Ok, let's wait to @michaelsauter

michaelsauter commented:
@adrianhurt @bjaglin Thanks for your great work on this. I love the idea in general, so please go ahead with it. I had a quick look through the code and found it a bit difficult to track, but it's quite a big change so that's to be expected. Since I don't have much time at the moment, I will leave this with you guys if you're okay with it. Once you think it's done, I can do a final walk-through and merge it.

adrianhurt commented:
Sorry to be disappeared for so much time. I'll try to finish this in a few days when I get some free time.

jleskovar commented:
Hi @adrianhurt 
Just wondering where you guys were at with the PR?
Cheers

michaelsauter commented:
Closing this for now due to inactivity. Feel free to reopen after 2.0 is released.

michaelsauter commented:
I'm not sure if you're still interested in this, but I loved the idea of this. Finally got around to implement it, basing it somewhat loosely on this PR.

The main commits are:
* 4e7c786136bac7dbd4f4aeb7fa68c2c912ec5470 extracts an interface with basic container information for use in templates
* 73b11071ae0e12e9484375c6309c88d20d14228a adds the actual command
